libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -1941,6 +1941,13 @@ void IRGeneratorForStatements::endVisit(MemberAccess const& _memberAccess)
 			ContractDefinition const& contract = contractType.contractDefinition();
 			define(_memberAccess) << formatNumber(u256{contract.interfaceId()} << (256 - 32)) << "\n";
 		}
+		else if (member == "typehash")
+		{
+			Type const* arg = dynamic_cast<MagicType const&>(*_memberAccess.expression().annotation().type).typeArgument();
+			auto const& structType = dynamic_cast<StructType const&>(*arg);
+			StructDefinition const& struct_ = structType.structDefinition();
```

Comment:
This conversion can fail. Can you add asserts here that `arg` is always a Struct definition.

This also needs to be gracefully handled in typechecker, i.e., the only argument for `type(X).typehash` should be when `X` is a struct (with additional conditions).

Reply:
Changed type reference to type pointer and added `solAssets`